### ISSUE TITLE
.gitignore is not working

### ISSUE THREAD

**Issue by PooyaHekmati** [RuminantFarmSystems/RuFaS#857]
There have been multiple cases where a file has been changed and synced, while it is mentioned in the gitignore. Investigate the problem.

Possibly related:
https://stackoverflow.com/questions/25436312/gitignore-not-working
https://gist.github.com/pavankjadda/2bb6fbdd8786e1f57fd7bcbcc666b51d
https://pavankjadda.dev/how-to-fix-gitignore-not-working-issue/

**Comment by Isgotkowitz** [comment]
@PooyaHekmati could you give an example of a file that this is happening with?

**Comment by PooyaHekmati** [comment]
> @PooyaHekmati could you give an example of a file that this is happening with?

An example is PR #751 deleted `output/output_filters/_json_all_variables.txt` while we had `output/output_filters/*` in the gitignore.

**Comment by ltp23** [comment]
I think one working solution for me I just found is to do things one directory level at a time. 

For example

```
# Exclude everything inside the output directory
output/*

# Re-include the output_filters directory
!output/output_filters/

# Exclude everything inside the output_filters directory
output/output_filters/*

# Re-include the needed file
!output/output_filters/json_test.txt
```

**Comment by Isgotkowitz** [comment]
I have tried the solution outlined in the links above. I think it has the desired effect but there are some important things to note about the behavior.

To expand, I tried this solution on branch_a and it made RUFAS/scenario_manager.py untracked without deleting it - which is great because now I can change it freely and not worry about it finding its way into my PRs! The only issue is that when I switch to another branch where I have not uncached everything, call it branch_b, RUFAS/scenario_manager.py disappears on branch_a. Once I clear the cache on branch_b though, I can switch back to branch_a, recreate RUFAS/scenario_manager.py, and switch between branch_a and branch_b with RUFAS/scenario_manager.py being kept.

In short, the linked solution works but requires clearing the cache on every branch you're working on and recreating any files that you have lost.

**Comment by ew3361zh** [comment]
> I have tried the solution outlined in the links above. I think it has the desired effect but there are some important things to note about the behavior.
> 
> To expand, I tried this solution on branch_a and it made RUFAS/scenario_manager.py untracked without deleting it - which is great because now I can change it freely and not worry about it finding its way into my PRs! The only issue is that when I switch to another branch where I have not uncached everything, call it branch_b, RUFAS/scenario_manager.py disappears on branch_a. Once I clear the cache on branch_b though, I can switch back to branch_a, recreate RUFAS/scenario_manager.py, and switch between branch_a and branch_b with RUFAS/scenario_manager.py being kept.
> 
> In short, the linked solution works but requires clearing the cache on every branch you're working on and recreating any files that you have lost.

@Isgotkowitz Would you be able to repeat this and add some screenshots to this issue? I'm having trouble understanding this fully.

**Comment by Isgotkowitz** [comment]
> @Isgotkowitz Would you be able to repeat this and add some screenshots to this issue? I'm having trouble understanding this fully.

@ew3361zh Yes - maybe a zoom call would be even better?

**Comment by ew3361zh** [comment]
After some conversations with Ed and some research, I think .gitignore is actually working as intended but that its intended functionality is not meeting and will not meet our needs. Adding a file or folder to the .gitignore file once the file/folder has already been added to/tracked by Git doesn't work. You have to remove it from being tracked. The main solution is using the `git rm -rf --cached` (linked above) which will remove the files/directories from being tracked locally but as soon as you push these changes to GitHub, it will tell GitHub to stop tracking these files as well so they will be removed from the remote repo. Changes to the file/folder will then not be tracked locally because it's listed in the .gitignore but then it will have been removed from GitHub and the only way to get it back to GitHub is to remove it from your .gitignore which will have it be tracked by git which means immediately adding it back to .gitignore won't have any effect without removing the tracking of this file from the cache...which will remove it from GitHub. A vicious cycle. We obviously don't want this.

Proposed alternative 1: the `git update-index --skip-worktree <file_path>` command. This seems like a great option initially because it tells the local git repo to stop tracking changes to whatever file/dir you specify but doesn't pass this on to GitHub when you push changes to the remote repo. Unfortunately, this is hyper local and seems like it will lead to some massive issues with frequently changed files like the json filters txt file or the scenario_manager.py file. There's no easy or clean way to make sure everyone is working from the exact version of each file we want to stop tracking locally and it would be cumbersome to add a new file to the list of files we didn't want tracked locally (and then make sure everyone has used the `no-skip-worktree` flag at the right time). 
For a good explanation of the problem, see: https://www.codeproject.com/Articles/5364391/Git-Problem-with-skip-worktree-and-Pull

Proposed alternative 2: I haven't fully explored the feasibility of this but had some initial discussions with @Isgotkowitz about using the `get-changed-file` action we use for flake8 linting and adding files we don't want changed at all to this list. The idea would be that we'd roll out not to change this list of files to the whole team but then add the files we don't want changed to the get-changed-file flake8 action and have it fail the changed-file check in GitHub if any of those files was altered and those changes pushed on a PR (regardless of whether they were violating flake8 guidelines). Perhaps we could even customize the fail message on those specific files (like a link to the file on main so that someone could easily copy the original file and paste it over their local changes to undo whatever they did in case they don't remember all the details). If it's not too difficult to add the files, I liked this option best because it gives us a centralized way to control which files are allowed to be changed.

Curious to know what you all think @Isgotkowitz @PooyaHekmati @ltp23 @allisterakun

**Comment by ew3361zh** [comment]
Something like this for having a GitHub action to monitor changes on files we don't want changed:


name: Warn on Unauthorized Changes

on:
  pull_request:
    paths:
      - 'path/to/restricted/file'

jobs:
  warn-on-changes:
    runs-on: ubuntu-latest

    steps:
    - name: Check for Unauthorized Changes
      id: check_changes
      run: |
        if git diff --name-only ${{ github.event.before }} ${{ github.sha }} | grep -q 'path/to/restricted/file'; then
          echo "::set-output name=changesDetected::true"
        else
          echo "::set-output name=changesDetected::false"
        fi

    - name: Comment on Pull Request
      if: steps.check_changes.outputs.changesDetected == 'true'
      run: |
        echo "Warning: You've made changes to 'path/to/restricted/file' which are not allowed."
        echo "Please review and remove these changes if they are not intended."

    - name: Set Status Check
      if: steps.check_changes.outputs.changesDetected == 'true'
      uses: actions/github-script@0.11.5
      with:
        github-token: ${{ secrets.GITHUB_TOKEN }}
        script: |
          github.checks.create({
            owner: context.repo.owner,
            repo: context.repo.repo,
            name: 'Unauthorized Changes',
            head_sha: context.sha,
            status: 'completed',
            conclusion: 'failure',
            output: {
              title: 'Unauthorized Changes Detected',
              summary: 'You have made unauthorized changes to the restricted file.'
            }
          })

**Comment by ew3361zh** [comment]
Closed with PR #911
